mac80211: restore ath9k functionality

neheb opened the pull request:
The initial ath9k_platform_data removal ended up breaking ath9k usage.

Revert everything and work backwards.

Now ath9k_platform.h is only used by 552-ath9k-ahb_of.patch. ath9k_platform.h goes away in 6.13, where the removal can happen by changing a single patch.

ping @kempniu is the LED still broken?

neheb committed:
mac80211: ath9k: clean up gpiochip

Simplify patch with gpiochip_add_data, struct reduction, new GPIO API,
and header cleanup.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17445
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb committed:
mac80211: ath9k: remove platform leds

These only work with and are useful with mach files. Now that those are
gone, this can go too.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17445
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb committed:
mac80211: ath9k: remove gpio buttons support

This is only used by mach files, which are no longer used in OpenWrt.

Allows removing a custon ath9k_platform.h file.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17445
Signed-off-by: Robert Marko <robimarko@gmail.com>

kempniu commented:
> ping @kempniu is the LED still broken?

With just this branch (657368aae0f713618fa1e25885ae9907899bca8f), yes, it is still broken on the Mikrotik RouterBOARD 951Ui-2HnD.  An expected error message is logged:

    [   20.301169] ath: EEPROM regdomain: 0x0
    [   20.301201] ath: EEPROM indicates default country code should be used
    [   20.301209] ath: doing EEPROM country->regdmn map search
    [   20.301227] ath: country maps to regdmn code: 0x3a
    [   20.301236] ath: Country alpha2 being used: US
    [   20.301245] ath: Regpair used: 0x3a
    [   20.314661] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
    [   20.315754] ath: phy0: request GPIO11 failed:-517
    [   20.320645] ------------[ cut here ]------------
    [   20.325461] WARNING: CPU: 0 PID: 570 at backports-6.12.6/drivers/net/wireless/ath/ath9k/hw.c:2920 ath_create_gpio_led+0x148/0x20c [ath9k]
    [   20.338389] Modules linked in: ath9k(O+) ath9k_common(O) pppoe ppp_async nft_fib_inet nf_flow_table_inet ath9k_hw(O) ath(O) pppox ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack mac80211(O) cfg80211(O) slhc rng_core nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_syslog nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c crc_ccitt compat(O) sha512_generic seqiv sha3_generic jitterentropy_rng drbg hmac geniv rng cmac ohci_platform ohci_hcd fsl_mph_dr_of ehci_platform ehci_fsl ehci_hcd gpio_button_hotplug(O) phy_ar7200_usb phy_ar7100_usb usbcore nls_base usb_common crc32c_generic
    [   20.409612] CPU: 0 PID: 570 Comm: kmodloader Tainted: G           O       6.6.68 #0
    [   20.417527] Stack : 00000000 0000000b 82c83afc 800c4ccc 00000000 00000004 00000000 00000000
    [   20.426200]         81e93ac4 80e60000 809d0000 80866720 81b74888 00000001 81e93a68 9b266383
    [   20.434863]         00000000 00000000 80866720 81e939a8 ffffefff 00000000 00000000 ffffffea
    [   20.443528]         000000a4 81e939b4 000000a4 80951760 00000001 80866720 00000000 81ec14d8
    [   20.452192]         00000009 0000000b 82c83afc 00000004 00000018 8047fbdc 00000000 80e60000
    [   20.460856]         ...
    [   20.463393] Call Trace:
    [   20.465919] [<800667d4>] show_stack+0x28/0xf0
    [   20.470456] [<80776ad8>] dump_stack_lvl+0x38/0x60
    [   20.475337] [<80084f90>] __warn+0xb0/0xe4
    [   20.479495] [<80085020>] warn_slowpath_fmt+0x5c/0xac
    [   20.484631] [<81ec14d8>] ath_create_gpio_led+0x148/0x20c [ath9k]
    [   20.490963] [<81ec17dc>] ath_init_leds+0xfc/0x1a0 [ath9k]
    [   20.496626] [<81ec2b30>] ath9k_init_device+0x8b8/0xa84 [ath9k]
    [   20.502736] [<81ed0374>] ath_pci_exit+0x464/0x57c [ath9k]
    [   20.508399] 
    [   20.509976] ---[ end trace 0000000000000000 ]---

That's because the GPIO base is still dynamically allocated since ee01666cd32e156214fe1ebc250c5c9ed7256aba, so a legacy `gpio_request_one()` call performed by the ath9k driver with an integer argument (11) fails.  Reverting ee01666cd32e156214fe1ebc250c5c9ed7256aba on top of 657368aae0f713618fa1e25885ae9907899bca8f makes the WLAN LED work again with the existing DTS file for this router.

Alternatively, applying the following [change](https://github.com/openwrt/openwrt/pull/12226#issuecomment-2228104892) on top of 657368aae0f713618fa1e25885ae9907899bca8f also makes the WLAN LED work again, with the [previously discussed](https://github.com/openwrt/openwrt/pull/15949#issuecomment-2250887694) caveat of leaving a useless LED device in sysfs:

```diff
diff --git a/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts b/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts
index 65003cb9a1..e61eed8ae5 100644
--- a/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts
+++ b/target/linux/ath79/dts/ar9344_mikrotik_routerboard-951ui-2hnd.dts
@@ -40,6 +40,12 @@
                        label = "green:port5";
                        gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
                };
+
+               wlan {
+                       label = "green:wlan";
+                       gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+                       linux,default-trigger = "phy0tpt";
+               };
        };
 
        gpio-export {
@@ -68,7 +74,3 @@
 &eth1 {
        compatible = "qca,ar9340-eth", "syscon";
 };
-
-&wmac {
-       qca,led-pin = /bits/ 8 <11>;
-};
```

hauke commented:
@robimarko  reverted some changes, please rebase this PR.

neheb commented:
@kempniu it’s an open question whether or not to use that solution. The alternative is to try those upstream patches. I say that even though there are two users of this feature.

neheb committed:
ath79: fix GPIO numbering with ath9k

ath9k base starts at 512. This is a problem as it uses gpio_request_one
to request the GPIO, which is legacy API.

This upstream pending patch needs to also be ported to mac80211.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17445
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb committed:
mac80211: fix led-pin usage

ath9k uses a deprecated GPIO API that assumes a starting base of 0. This
is no longer the case and so must be fixed.

Upstream pending patch.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17445
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb commented:
@kempniu I added that v2 patch from upstream. This might fix qca,led-pin.

Neustradamus commented:
It replaces:
- https://github.com/openwrt/openwrt/pull/16635
- https://github.com/openwrt/openwrt/pull/17402

kempniu commented:
I'm getting a kernel crash on boot with b4020572ab96ad9cded5d7029b14eecb1627ff98:

    OpenWrt kernel loader for AR7XXX/AR9XXX
    Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
    Decompressing kernel... done!
    Starting kernel at 80060000...
    
    [    0.000000] Linux version 6.6.68 (openwrt@openwrt) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 13.3.0 r0+28473-b4020572ab) 13.3.0, GNU ld (GNU Binutils) 2.42) #0 Thu Jan2 00:21:44 2025
    [    0.000000] printk: bootconsole [early0] enabled
    [    0.000000] CPU0 revision is: 0001974c (MIPS 74Kc)
    [    0.000000] MIPS: machine is Mikrotik RouterBOARD 951Ui-2HnD
    [    0.000000] SoC: Atheros AR9344 rev 2
    [    0.000000] Initrd not found or empty - disabling initrd
    [    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
    [    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
    [    0.000000] Zone ranges:
    [    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
    [    0.000000] Movable zone start for each node
    [    0.000000] Early memory node ranges
    [    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
    [    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
    [    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
    [    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
    [    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
    [    0.000000] Writing ErrCtl register=00000000
    [    0.000000] Readback ErrCtl register=00000000
    [    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
    [    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
    [    0.000000] Memory: 115176K/131072K available (7454K kernel code, 593K rwdata, 1636K rodata, 4644K init, 229K bss, 15896K reserved, 0K cma-reserved)
    [    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
    [    0.000000] NR_IRQS: 51
    [    0.000000] CPU clock: 600.000 MHz
    [    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6370868154 ns
    [    0.000001] sched_clock: 32 bits at 300MHz, resolution 3ns, wraps every 7158278654ns
    [    0.008947] Calibrating delay loop... 299.82 BogoMIPS (lpj=1499136)
    [    0.086042] pid_max: default: 32768 minimum: 301
    [    0.103432] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
    [    0.111765] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
    [    0.131050] RCU Tasks Trace: Setting shift to 0 and lim to 1 rcu_task_cb_adjust=1 rcu_task_cpu_ids=1.
    [    0.146019] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
    [    0.157256] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
    [    0.168203] pinctrl core: initialized pinctrl subsystem
    [    0.177811] NET: Registered PF_NETLINK/PF_ROUTE protocol family
    [    0.187668] /ahb/apb: Fixed dependency cycle(s) with /ahb/apb/interrupt-controller@18060010
    [    0.213829] clocksource: Switched to clocksource MIPS
    [    0.232799] NET: Registered PF_INET protocol family
    [    0.238653] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
    [    0.247898] tcp_listen_portaddr_hash hash table entries: 1024 (order: 0, 4096 bytes, linear)
    [    0.257572] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
    [    0.266388] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
    [    0.275107] TCP bind hash table entries: 1024 (order: 1, 8192 bytes, linear)
    [    0.283144] TCP: Hash tables configured (established 1024 bind 1024)
    [    0.291173] MPTCP token hash table entries: 512 (order: 0, 6144 bytes, linear)
    [    0.299769] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
    [    0.307246] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
    [    0.316032] NET: Registered PF_UNIX/PF_LOCAL protocol family
    [    0.322522] PCI: CLS 0 bytes, default 32
    [    0.896496] workingset: timestamp_bits=14 max_order=15 bucket_order=1
    [    0.905664] squashfs: version 4.0 (2009/01/31) Phillip Lougher
    [    0.912302] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
    [    0.932615] pinctrl-single 1804002c.pinmux: 544 pins, size 68
    [    0.940083] gpio-export gpio-export: 2 gpio(s) exported
    [    0.947529] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
    [    0.956136] printk: console [ttyS0] disabled
    [    0.961631] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 11, base_baud = 1562500) is a 16550A
    [    0.971611] printk: console [ttyS0] enabled
    [    0.971611] printk: console [ttyS0] enabled
    [    0.980689] printk: bootconsole [early0] disabled
    [    0.980689] printk: bootconsole [early0] disabled
    [    1.000962] nand: device found, Manufacturer ID: 0xec, Chip ID: 0xf1
    [    1.007597] nand: Samsung NAND 128MiB 3,3V 8-bit
    [    1.012371] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
    [    1.020350] Scanning device for bad blocks
    [    3.666218] 3 fixed-partitions partitions found on MTD device ar934x-nand
    [    3.673265] Creating 3 MTD partitions on "ar934x-nand":
    [    3.678731] 0x000000000000-0x000000040000 : "booter"
    [    3.686000] 0x000000040000-0x000000400000 : "kernel"
    [    4.019428] 0x000000400000-0x000008000000 : "ubi"
    [    4.032192] spi-nor spi0.0: mx25l512e (64 Kbytes)
    [    4.037387] 1 routerbootpart partitions found on MTD device spi0.0
    [    4.043788] Creating 1 MTD partitions on "spi0.0":
    [    4.048773] 0x000000000000-0x000000010000 : "partitions"
    [    4.056646] 4 routerbootpart partitions found on MTD device partitions
    [    4.063420] Creating 4 MTD partitions on "partitions":
    [    4.068785] 0x000000000000-0x00000000c000 : "routerboot"
    [    4.075157] 0x00000000c000-0x00000000d000 : "hard_config"
    [    4.081570] 0x00000000d000-0x00000000e000 : "bios"
    [    4.087346] 0x00000000e000-0x00000000f000 : "soft_config"
    [    4.099418] ag71xx-legacy 19000000.eth: invalid MAC address, using random address
    [    5.031482] ag71xx-legacy 19000000.eth: Could not connect to PHY device. Deferring probe.
    [    5.040333] ag71xx-legacy 1a000000.eth: invalid MAC address, using random address
    [    6.696224] switch0: Atheros AR8229 rev. 1 switch registered on mdio.0
    [    6.781900] ag71xx-legacy 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
    [    6.792632] eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
    [    6.799643] ath79-wdt 18060008.wdt: timeout value must be 0 < timeout < 10, using 10
    [    6.812219] MikroTik RouterBOARD hardware configuration sysfs driver v0.08
    [    6.821703] MikroTik RouterBOARD software configuration sysfs driver v0.05
    [    6.830703] NET: Registered PF_INET6 protocol family
    [    6.847483] Segment Routing with IPv6
    [    6.851384] In-situ OAM (IOAM) with IPv6
    [    6.855682] NET: Registered PF_PACKET protocol family
    [    6.861318] 8021q: 802.1Q VLAN Support v1.8
    [    6.911269] ag71xx-legacy 19000000.eth: invalid MAC address, using random address
    [    7.335029] ag71xx-legacy 19000000.eth: connected to PHY at mdio.0:1f:04 [uid=004dd042, driver=Qualcomm Atheros QCA9561 built-in PHY]
    [    7.348558] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
    [    7.358012] UBI: auto-attach mtd2
    [    7.361486] ubi0: attaching mtd2
    [    8.620262] ubi0: scanning is finished
    [    8.642640] ubi0: attached mtd2 (name "ubi", size 124 MiB)
    [    8.648394] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 129024 bytes
    [    8.655512] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 512
    [    8.662437] ubi0: VID header offset: 512 (aligned 512), data offset: 2048
    [    8.669466] ubi0: good PEBs: 992, bad PEBs: 0, corrupted PEBs: 0
    [    8.675681] ubi0: user volume: 2, internal volumes: 1, max. volumes count: 128
    [    8.683144] ubi0: max/mean erase counter: 4/1, WL threshold: 4096, image sequence number: 641300361
    [    8.692505] ubi0: available PEBs: 0, total reserved PEBs: 992, PEBs reserved for bad PEB handling: 20
    [    8.702089] ubi0: background thread "ubi_bgt0d" started, PID 267
    [    8.709841] block ubiblock0_0: created from ubi0:0(rootfs)
    [    8.715598] ubiblock: device ubiblock0_0 (rootfs) set to be root filesystem
    [    8.723019] clk: Disabling unused clocks
    [    8.749518] Freeing unused kernel image (initmem) memory: 4644K
    [    8.755691] This architecture does not have kernel memory protection.
    [    8.762359] Run /init as init process
    [    9.333314] init: Console is alive
    [    9.337405] init: - watchdog -
    [    9.365447] kmodloader: loading kernel modules from /etc/modules-boot.d/*
    [    9.391637] usbcore: registered new interface driver usbfs
    [    9.397493] usbcore: registered new interface driver hub
    [    9.403086] usbcore: registered new device driver usb
    [    9.418996] gpio_button_hotplug: loading out-of-tree module taints kernel.
    [    9.439106] ehci-platform 1b000000.usb: EHCI Host Controller
    [    9.445065] ehci-platform 1b000000.usb: new USB bus registered, assigned bus number 1
    [    9.453305] ehci-platform 1b000000.usb: irq 3, io mem 0x1b000000
    [    9.483851] ehci-platform 1b000000.usb: USB 2.0 started, EHCI 1.00
    [    9.491822] hub 1-0:1.0: USB hub found
    [    9.496413] hub 1-0:1.0: 1 port detected
    [    9.513021] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
    [    9.522953] init: - preinit -
    [   10.523863] random: crng init done
    Cannot parse config file '/etc/fw_env.config': No such file or directory
    Failed to find NVMEM device
    [   11.838331] eth0: link up (1000Mbps/Full duplex)
    Press the [f] key and hit [enter] to enter failsafe mode
    Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
    [   16.075328] eth0: link down
    [   16.104288] procd: - early -
    [   16.107623] procd: - watchdog -
    [   16.723245] procd: - watchdog -
    [   16.726958] procd: - ubus -
    [   16.787791] procd: - init -
    Please press Enter to activate this console.
    [   17.797592] kmodloader: loading kernel modules from /etc/modules.d/*
    [   18.538626] jitterentropy: Initialization failed with host not compliant with requirements: 9
    [   18.931664] Loading modules backported from Linux version v6.12.6-0-ge9d65b48ce1a
    [   18.939467] Backport generated by backports.git v6.1.110-1-35-g410656ef04d2
    [   19.278867] PPP generic driver version 2.4.2
    [   19.296972] NET: Registered PF_PPPOX protocol family
    [   19.404308] CPU 0 Unable to handle kernel paging request at virtual address 00001818, epc == 81ed027c, ra == 81ed0270
    [   19.415342] Oops[#1]:
    [   19.417694] CPU: 0 PID: 569 Comm: kmodloader Tainted: G           O       6.6.68 #0
    [   19.425611] $ 0   : 00000000 00000001 0000000b 80001f21
    [   19.431031] $ 4   : 0000000b 82cfbc6c 00000000 00000001
    [   19.436452] $ 8   : 00000091 81be16ea 00000010 ffffff84
    [   19.441872] $12   : 82c1001c 82cfbbba ffffffff fffffffb
    [   19.447293] $16   : 0000000c 818a2400 b8100000 80950000
    [   19.452714] $20   : 82c104c0 82c14ce8 82c116e0 818a2410
    [   19.458135] $24   : 00000010 00000020
    [   19.463556] $28   : 82cfa000 82cfbc50 00000000 81ed0270
    [   19.468976] Hi    : 00000008
    [   19.471951] Lo    : 00000004
    [   19.474926] epc   : 81ed027c ath_pci_exit+0x38c/0x500 [ath9k]
    [   19.481010] ra    : 81ed0270 ath_pci_exit+0x380/0x500 [ath9k]
    [   19.487004] Status: 1100fc03 KERNEL EXL IE
    [   19.491341] Cause : 0080000c (ExcCode 03)
    [   19.495481] BadVA : 00001818
    [   19.498455] PrId  : 0001974c (MIPS 74Kc)
    [   19.502506] Modules linked in: ath9k(O+) ath9k_common(O) pppoe ppp_async nft_fib_inet nf_flow_table_inet ath9k_hw(O) ath(O) pppox ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack mac80211(O) cfg80211(O) slhc rng_core nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_syslog nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c crc_ccitt compat(O) sha512_generic seqiv sha3_generic drbg hmac geniv rng cmac ohci_platform ohci_hcd fsl_mph_dr_of ehci_platform ehci_fsl ehci_hcd gpio_button_hotplug(O) phy_ar7200_usb phy_ar7100_usb usbcore nls_base usb_common crc32c_generic
    [   19.572079] Process kmodloader (pid: 569, threadinfo=86d55609, task=f8c7feb7, tls=77e97df4)
    [   19.580717] Stack : 0002434c 0002434c 818a1600 8028d4fc 00000000 82c116e0 80414c0b80002fe4
    [   19.589390]         81bcfae0 81bcfc60 818a1600 00000001 00000008 8028fd58 82e44a8081f4bc80
    [   19.598055]         00000001 00000000 00000000 00000004 808a3b04 818a1600 81bcfc60e9bbfd4b
    [   19.606718]         818a2410 81c382a8 81c382a8 00000000 00000008 0002434c 0002434c00010000
    [   19.615382]         00010001 804968ac 81c382a8 00000000 00000008 0002434c 0002434c818a2410
    [   19.624047]         ...
    [   19.626585] Call Trace:
    [   19.629110] [<81ed027c>] ath_pci_exit+0x38c/0x500 [ath9k]
    [   19.634792] [<81ed0270>] ath_pci_exit+0x380/0x500 [ath9k]
    [   19.640452]
    [   19.641994] Code: 04400002  2402ffff  93a2001b <afc21818> 3c0582c1  8e2400cc  00003025  0c14d6cb  24a56434
    [   19.652110]
    [   19.653695] ---[ end trace 0000000000000000 ]---
    [   19.658478] Kernel panic - not syncing: Fatal exception
    [   19.663879] Rebooting in 1 seconds..

kempniu commented:
8defcad111cbe9fdbb679e46b8c4f6c7549b0100 indeed boots fine, thanks, but the GPIO request fails with `-ENOENT`:

    [   20.340977] ath: EEPROM regdomain: 0x0
    [   20.341008] ath: EEPROM indicates default country code should be used
    [   20.341016] ath: doing EEPROM country->regdmn map search
    [   20.341034] ath: country maps to regdmn code: 0x3a
    [   20.341044] ath: Country alpha2 being used: US
    [   20.341052] ath: Regpair used: 0x3a
    [   20.354462] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
    [   20.355558] ath: phy0: request GPIO11 failed:-2
    [   20.360268] ------------[ cut here ]------------
    [   20.365093] WARNING: CPU: 0 PID: 570 at backports-6.12.6/drivers/net/wireless/ath/ath9k/hw.c:2923 ath_create_gpio_led+0x148/0x20c [ath9k]
    [   20.378019] Modules linked in: ath9k(O+) ath9k_common(O) pppoe ppp_async nft_fib_inet nf_flow_table_inet ath9k_hw(O) ath(O) pppox ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_ct nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack mac80211(O) cfg80211(O) slhc rng_core nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_log_syslog nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c crc_ccitt compat(O) sha512_generic seqiv sha3_generic jitterentropy_rng drbg hmac geniv rng cmac ohci_platform ohci_hcd fsl_mph_dr_of ehci_platform ehci_fsl ehci_hcd gpio_button_hotplug(O) phy_ar7200_usb phy_ar7100_usb usbcore nls_base usb_common crc32c_generic
    [   20.449236] CPU: 0 PID: 570 Comm: kmodloader Tainted: G           O       6.6.68 #0
    [   20.457149] Stack : 00000000 0000000b 81b318fc 800c4ccc 00000000 00000004 00000000 00000000
    [   20.465821]         82e67ac4 80e60000 809d0000 80866720 81bdbd08 00000001 82e67a68 d0996231
    [   20.474485]         00000000 00000000 80866720 82e679a8 ffffefff 00000000 00000000 ffffffea
    [   20.483149]         000000a4 82e679b4 000000a4 80951760 00000001 80866720 00000000 81ec14d8
    [   20.491813]         00000009 0000000b 81b318fc 00000004 00000018 8047fc44 00000000 80e60000
    [   20.500478]         ...
    [   20.503014] Call Trace:
    [   20.505540] [<800667d4>] show_stack+0x28/0xf0
    [   20.510077] [<80776b40>] dump_stack_lvl+0x38/0x60
    [   20.514959] [<80084f90>] __warn+0xb0/0xe4
    [   20.519117] [<80085020>] warn_slowpath_fmt+0x5c/0xac
    [   20.524252] [<81ec14d8>] ath_create_gpio_led+0x148/0x20c [ath9k]
    [   20.530584] [<81ec17dc>] ath_init_leds+0xfc/0x1a0 [ath9k]
    [   20.536237] [<81ec2b30>] ath9k_init_device+0x8b8/0xa84 [ath9k]
    [   20.542339] [<81ed0374>] ath_pci_exit+0x464/0x57c [ath9k]
    [   20.547994] 
    [   20.549569] ---[ end trace 0000000000000000 ]---

neheb commented:
ping @linusw 

Alright so I'm going to abandon trying to fix qca,led-pin on ath79. Maybe someday it will work. It still works on mpc85xx so I'll keep it around.

neheb commented:
@hauke this is in good shape now.

kempniu commented:
I had a longer while to take a closer look at the patch in 8ffcf5c80d9b0381126e2780e2533c9d98c425c5 and what it is supposed to achieve.  I found two issues with it that prevent it from working as desired.

First, the lookup table device ID mismatch: it is [set](https://github.com/openwrt/openwrt/blob/8ffcf5c80d9b0381126e2780e2533c9d98c425c5/target/linux/ath79/patches-6.6/341-wifi-ath9k-obtain-system-gpios.patch#L207) to `ath9k` in the GPIO driver, but when the wireless driver requests a GPIO line, `ah->dev` is [passed](https://github.com/openwrt/openwrt/blob/8ffcf5c80d9b0381126e2780e2533c9d98c425c5/target/linux/ath79/patches-6.6/341-wifi-ath9k-obtain-system-gpios.patch#L262) as the first argument to `devm_gpiod_get_index()`, which causes `gpiod_find_lookup_table()` to use the wireless device name (`18100000.wmac` in the case of the Mikrotik RouterBOARD 951Ui-2HnD) for matching against the lookup table's device ID.  That's where the `-ENOENT` comes from.

I don't know the relevant bits of code well enough to suggest a clean solution - obviously the ID simply needs to match, but when using the resource-managed version of `gpiod_get_index()` (i.e. `devm_gpiod_get_index()`), there is no possibility to pass a different device identifier (`ath9k` would work) and there is no way (?) for the GPIO driver to guess the name of the device that will request GPIOs from it (and that name is not fixed as it varies across platforms).

One ugly workaround that seems to work for me is to resort to the `NULL`-matching fallback in `gpiod_find_lookup_table()`, which enables a table with `dev_id` set to `NULL` to be matched for a `gpiod_get_index()` call with `dev` also set to `NULL`.  This requires manually releasing the GPIO descriptor when the wireless driver is done with it (because we're losing the benefits of using resource-managed functions), but that's not a big issue, I think - a bigger concern for me is that this change feels dangerous.  Here it is anyway:

```diff
diff --git a/drivers/gpio/gpio-ath79.c b/drivers/gpio/gpio-ath79.c
index 24ae18de730d..077c056709e3 100644
--- a/drivers/gpio/gpio-ath79.c
+++ b/drivers/gpio/gpio-ath79.c
@@ -243,8 +243,6 @@ static int ath79_gpio_register_wifi_descriptors(struct device *dev,
 	if (!lookup)
 		return -ENOMEM;
 
-	lookup->dev_id = "ath9k";
-
 	for (i = 0; i < ATH79K_WIFI_DESCS; i++) {
 		lookup->table[i] = (struct gpiod_lookup)
 			GPIO_LOOKUP_IDX(label, 0, NULL, i,
diff --git a/drivers/net/wireless/ath/ath9k/hw.c b/drivers/net/wireless/ath/ath9k/hw.c
index ee6705836746..e760350ea6d5 100644
--- a/drivers/net/wireless/ath/ath9k/hw.c
+++ b/drivers/net/wireless/ath/ath9k/hw.c
@@ -2735,7 +2735,7 @@ static void ath9k_hw_gpio_cfg_soc(struct ath_hw *ah, u32 gpio, bool out,
 		return;
 
 	/* Obtains a system specific GPIO descriptor from another GPIO controller */
-	gpiod = devm_gpiod_get_index(ah->dev, NULL, gpio, flags);
+	gpiod = gpiod_get_index(NULL, NULL, gpio, flags);
 
 	if (IS_ERR(gpiod)) {
 		err = PTR_ERR(gpiod);
@@ -2806,6 +2806,11 @@ void ath9k_hw_gpio_free(struct ath_hw *ah, u32 gpio)
 		return;
 
 	WARN_ON(gpio >= ah->caps.num_gpio_pins);
+
+	if (ah->gpiods[gpio]) {
+		gpiod_put(ah->gpiods[gpio]);
+		ah->gpiods[gpio] = NULL;
+	}
 }
 EXPORT_SYMBOL(ath9k_hw_gpio_free);
 
```

However, I mentioned that there are two issues with the patch and this is where we get to the second one: even with the lookup table matching fixed, in `ath79_gpio_register_wifi_descriptors()`, the `chip_hwnum` member of *all* the `struct gpiod_lookup` structures created in the lookup table is [set](https://github.com/openwrt/openwrt/blob/8ffcf5c80d9b0381126e2780e2533c9d98c425c5/target/linux/ath79/patches-6.6/341-wifi-ath9k-obtain-system-gpios.patch#L211) to `0` (it's the second argument of the `GPIO_LOOKUP_IDX()` macro).  This causes all 32 GPIOs that the wireless driver could potentially request to be erroneously mapped to the GPIO chip's first line.  The fix for this particular issue seems to be simple for a change:

```diff
diff --git a/drivers/gpio/gpio-ath79.c b/drivers/gpio/gpio-ath79.c
index 077c056709e3..4a3f514c55bf 100644
--- a/drivers/gpio/gpio-ath79.c
+++ b/drivers/gpio/gpio-ath79.c
@@ -245,7 +245,7 @@ static int ath79_gpio_register_wifi_descriptors(struct device *dev,
 
 	for (i = 0; i < ATH79K_WIFI_DESCS; i++) {
 		lookup->table[i] = (struct gpiod_lookup)
-			GPIO_LOOKUP_IDX(label, 0, NULL, i,
+			GPIO_LOOKUP_IDX(label, i, NULL, i,
 					GPIO_ACTIVE_HIGH);
 	}
 
```

With these two patches applied, `qca,led-pin` works as desired on my Mikrotik RouterBOARD 951Ui-2HnD.  I do have concerns about the ~~horrible hack~~workaround that I came up with for the first issue, but I'm sure the author of the upstream patch would be able to fix it properly :-)

kempniu commented:
Okay, leaving the lookup table's `dev_id` set to `NULL`, but setting `con_id` in each lookup table entry, combined with using plain `gpiod_get_index()` with a matching `con_id` string seems to work and should be safer than the original hack I proposed above.  Applying these tweaks on top of the upstream patch makes the WLAN LED work for me:

```diff
diff --git a/drivers/gpio/gpio-ath79.c b/drivers/gpio/gpio-ath79.c
index 24ae18de730d..7ee975907d80 100644
--- a/drivers/gpio/gpio-ath79.c
+++ b/drivers/gpio/gpio-ath79.c
@@ -243,11 +243,9 @@ static int ath79_gpio_register_wifi_descriptors(struct device *dev,
 	if (!lookup)
 		return -ENOMEM;
 
-	lookup->dev_id = "ath9k";
-
 	for (i = 0; i < ATH79K_WIFI_DESCS; i++) {
 		lookup->table[i] = (struct gpiod_lookup)
-			GPIO_LOOKUP_IDX(label, 0, NULL, i,
+			GPIO_LOOKUP_IDX(label, i, "ath9k", i,
 					GPIO_ACTIVE_HIGH);
 	}
 
diff --git a/drivers/net/wireless/ath/ath9k/hw.c b/drivers/net/wireless/ath/ath9k/hw.c
index ee6705836746..11b2fe861a6c 100644
--- a/drivers/net/wireless/ath/ath9k/hw.c
+++ b/drivers/net/wireless/ath/ath9k/hw.c
@@ -2735,7 +2735,7 @@ static void ath9k_hw_gpio_cfg_soc(struct ath_hw *ah, u32 gpio, bool out,
 		return;
 
 	/* Obtains a system specific GPIO descriptor from another GPIO controller */
-	gpiod = devm_gpiod_get_index(ah->dev, NULL, gpio, flags);
+	gpiod = gpiod_get_index(NULL, "ath9k", gpio, flags);
 
 	if (IS_ERR(gpiod)) {
 		err = PTR_ERR(gpiod);
@@ -2806,6 +2806,11 @@ void ath9k_hw_gpio_free(struct ath_hw *ah, u32 gpio)
 		return;
 
 	WARN_ON(gpio >= ah->caps.num_gpio_pins);
+
+	if (ah->gpiods[gpio]) {
+		gpiod_put(ah->gpiods[gpio]);
+		ah->gpiods[gpio] = NULL;
+	}
 }
 EXPORT_SYMBOL(ath9k_hw_gpio_free);
 
```

I don't know if this is "proper" from the upstream's standpoint, but perhaps it is an acceptable solution for the time being?

neheb commented:
Should I even bother fixing qca,led-pin when gpio-leds exists? ping @DragonBluep

neheb committed:
mac80211: ath9k: simplify probe with devm

Upstream backport.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17445
Signed-off-by: Robert Marko <robimarko@gmail.com>

kempniu commented:
Thanks.  557b31c37b8f01a06e75211fee4f7626358d3427 is still missing the [second fix](https://github.com/openwrt/openwrt/pull/17445#issuecomment-2570480649) for `GPIO_LOOKUP_IDX()` (the one for `chip_hwnum`).  Here's the patch for the patch:

```diff
diff --git a/target/linux/ath79/patches-6.6/341-wifi-ath9k-obtain-system-gpios.patch b/target/linux/ath79/patches-6.6/341-wifi-ath9k-obtain-system-gpios.patch
index 663058a57c..892b1990ed 100644
--- a/target/linux/ath79/patches-6.6/341-wifi-ath9k-obtain-system-gpios.patch
+++ b/target/linux/ath79/patches-6.6/341-wifi-ath9k-obtain-system-gpios.patch
@@ -206,7 +206,7 @@ Best regards,
 +
 +      for (i = 0; i < ATH79K_WIFI_DESCS; i++) {
 +              lookup->table[i] = (struct gpiod_lookup)
-+                      GPIO_LOOKUP_IDX(label, 0, "ath9k", i,
++                      GPIO_LOOKUP_IDX(label, i, "ath9k", i,
 +                                      GPIO_ACTIVE_HIGH);
 +      }
 +
```

neheb commented:
Ah missed that. Will fix later today.

kempniu commented:
Thanks, 8419ece89cc6089a8ecae7cc7bc2898561f1728c also makes the WLAN LED work fine for me :+1:

Just a couple of final comments:

  - c487b1f4fb3ebc06d7657288fca832c172a599b6:
      - the version of the patch included in this commit still uses the original `devm_gpiod_get_index()` call, which does not work; however, since it is properly patched in the mac80211 package by b8e1d18cb2a6cfd05a8e07206028ba20e0281179, it all works out in the end, at least on my router; what I'm trying to say is that I don't understand the rules for patching things in the kernel itself vs. in the mac80211 package, but I know you do, so you might want to double-check that c487b1f4fb3ebc06d7657288fca832c172a599b6 does the right thing,
      - you might want to consider adding a `Related: commit ee01666cd3 ("ath79: suppress GPIO static base allocation warning")` or similar tag as a cross-reference (`Fixes` feels a bit inadequate here as that commit did not do anything wrong per se IMHO),
      - feel free to add a `Tested-by: Michał Kępień <openwrt@kempniu.pl>` tag if you want to,
  - b8e1d18cb2a6cfd05a8e07206028ba20e0281179:
      - feel free to add a `Tested-by: Michał Kępień <openwrt@kempniu.pl>` tag if you want to.

I also submitted my feedback on the patch upstream to hopefully limit OpenWRT overhead for maintaining it in the long run:

https://patchwork.ozlabs.org/project/linux-gpio/patch/20240423-descriptors-wireless-v2-1-6d1d03b30bfa@linaro.org/#3438851

neheb commented:
The ath79 patch patches ath9k in the kernel but that does not get used. So it ends up being noop. I’ll remove it anyway.

neheb commented:
Removed. This is ready.

neheb commented:
Anything missing?

neheb commented:
ping

neheb commented:
ping

neheb commented:
anything missing?

robimarko commented:
I would say this has been cooking long enough

kempniu commented:
> I also submitted my feedback on the patch upstream to hopefully limit OpenWRT overhead for maintaining it in the long run:
> 
> https://patchwork.ozlabs.org/project/linux-gpio/patch/20240423-descriptors-wireless-v2-1-6d1d03b30bfa@linaro.org/#3438851

Just a heads-up that v6 of this patch (added to OpenWRT in 0ac79009dd3433d5e92a8311beae1041f39c3b79) has been accepted upstream and is currently queued for version 7.2 of the Linux kernel:

  - https://git.kernel.org/pub/scm/linux/kernel/git/wireless/wireless-next.git/commit/?id=34a5329beee86a22a446e27eb37f06caa63479ca
  - https://patchwork.ozlabs.org/project/linux-gpio/patch/20260317-descriptors-wireless-v6-1-b19ecff9cd2b@kernel.org/

It works fine on my hardware, though obviously the `#if IS_ENABLED(CONFIG_ATH9K_AHB)` guard needs to be removed beforehand due to the way things are currently built for OpenWRT.

neheb commented:
Great.
